twilio-audio-router/src/main/java/com/twilio/audio_manager/AudioDeviceSelector.java

```diff
@@ -0,0 +1,376 @@
+                enableSpeakerphone(savedSpeakerphoneEnabled);
+
+                audioManager.abandonAudioFocus(null);
+                state = State.STARTED;
+                break;
+            case STARTED:
+            case STOPPED:
+                // no-op;
+                break;
+        }
+    }
+
+    /**
+     * Select the desired {@link AudioDevice}. If the provided {@link AudioDevice} is not available
+     * no changes are made. If the provided {@link AudioDevice} is null an {@link AudioDevice} is
+     * chosen based on the following preference: Bluetooth, Wired Headset, Microphone, Speakerphone
+     *
+     * @param audioDevice The {@link AudioDevice} to use
+     */
+    public void selectDevice(@Nullable AudioDevice audioDevice) {
+        ThreadUtils.checkIsOnMainThread();
+        userSelectedDevice = audioDevice;
+        enumerateDevices();
+    }
+
+    /** @return The selected {@link AudioDevice} */
+    public @Nullable AudioDevice getSelectedAudioDevice() {
+        ThreadUtils.checkIsOnMainThread();
+        return selectedDevice != null
+                ? new AudioDevice(selectedDevice.type, selectedDevice.name)
+                : null;
+    }
+
+    public @NonNull List<AudioDevice> getAudioDevices() {
+        ThreadUtils.checkIsOnMainThread();
+        return Collections.unmodifiableList(new ArrayList<>(availableAudioDevices));
+    }
+
+    private void setAudioFocus() {
+        // Request audio focus before making any device switch.
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+            AudioAttributes playbackAttributes =
+                    new AudioAttributes.Builder()
+                            .setUsage(AudioAttributes.USAGE_VOICE_COMMUNICATION)
+                            .setContentType(AudioAttributes.CONTENT_TYPE_SPEECH)
+                            .build();
+            AudioFocusRequest focusRequest =
+                    new AudioFocusRequest.Builder(AudioManager.AUDIOFOCUS_GAIN_TRANSIENT)
+                            .setAudioAttributes(playbackAttributes)
+                            .setAcceptsDelayedFocusGain(true)
+                            .setOnAudioFocusChangeListener(i -> {})
+                            .build();
+            audioManager.requestAudioFocus(focusRequest);
+        } else {
+            audioManager.requestAudioFocus(
+                    focusChange -> {},
+                    AudioManager.STREAM_VOICE_CALL,
+                    AudioManager.AUDIOFOCUS_GAIN_TRANSIENT);
+        }
+        /*
+         * Start by setting MODE_IN_COMMUNICATION as default audio mode. It is
+         * required to be in this mode when playout and/or recording starts for
+         * best possible VoIP performance. Some devices have difficulties with speaker mode
+         * if this is not set.
+         */
+        audioManager.setMode(AudioManager.MODE_IN_COMMUNICATION);
+    }
+
+    private class WiredHeadsetReceiver extends BroadcastReceiver {
```

Comment:
I think this should be moved to a separate class. It would make the class easier to follow. We can give it an internal access modifier versus public after converting to Kotlin.